internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	return &ParsedSignedData{
+		Content:      signedData.EncapsulatedContentInfo.Content,
+		ContentType:  signedData.EncapsulatedContentInfo.ContentType,
+		Certificates: certs,
+		CRLs:         signedData.CRLs,
+		Signers:      signedData.SignerInfos,
+	}, nil
+}
+
+// Verify attempts to verify the content in the parsed signed data against the signer
+// information. The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed data.
+// If more than one signature is present, the successful validation of any signature
+// implies that the content in the parsed signed data is valid.
+// On successful verification, the list of signers that successfully verify is returned.
+// If all signatures fail to verify, the last error is returned.
+// References:
+// - RFC 5652 5   Signed-data Content Type
+// - RFC 5652 5.4 Message Digest Calculation Process
+// - RFC 5652 5.6 Signature Verification Process
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) Verify(opts x509.VerifyOptions) ([]SignerInfo, error) {
+	if len(d.Signers) == 0 {
+		return nil, ErrSignerNotFound
+	}
+	if len(d.Certificates) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	var verifiedSigners []SignerInfo
+	var lastErr error
+	for _, signer := range d.Signers {
+		if err := d.verify(signer, opts); err != nil {
+			lastErr = err
+		} else {
+			verifiedSigners = append(verifiedSigners, signer)
+		}
+	}
+	if len(verifiedSigners) == 0 {
+		return nil, lastErr
+	}
+	return verifiedSigners, nil
+}
+
+// verify verifies the trust in a top-down manner.
+// References:
+// - RFC 5652 5.4 Message Digest Calculation Process
+// - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verify(signer SignerInfo, opts x509.VerifyOptions) error {
+	// find signer certificate
+	cert := d.getCertificate(signer.SignerIdentifier)
+	if cert == nil {
+		return ErrCertificateNotFound
+	}
+
+	// verify signer certificate
+	if _, err := cert.Verify(opts); err != nil {
+		return VerificationError{Detail: err}
+	}
+
+	// verify signature
+	algorithm := getSignatureAlgorithmFromOID(
+		signer.DigestAlgorithm.Algorithm,
+		signer.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+	signed := d.Content
+	if len(signer.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signer.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+	if err := cert.CheckSignature(algorithm, signed, signer.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+
+	// verify attributes if present
```

Comment:
verify attributes are present

Reply:
We only verify attributes if they are present.